### PR TITLE
[Enhancement](Status) use `Status` to expose the error info more explicitly in `FlushToken`

### PR DESCRIPTION
**Pull request by bobhan1**
## Proposed changes
Currently, `FlushToken` only store the error code if `_do_flush_memtable` failed and it's hard to find the root cause due to the abuse of `InternalError`. This let  `FlushToken` store `Status` directly returned by `_do_flush_memtable` instead.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by bobhan1** [comment]
run buildall

File: be/src/olap/memtable_flush_executor.cpp [hunk 1]
```diff
@@ -127,13 +138,16 @@ void FlushToken::_flush_memtable(MemTable* memtable, int32_t segment_id, int64_t
...
+        if (!_flush_status.ok()) {
+            return;
+        }
     }
-    if (_flush_status.load() != OK) {
-        return;
+    if (!s.ok()) {
+        std::lock_guard wrlk(_flush_status_lock);
+        LOG(WARNING) << "Flush memtable failed with res = " << s;
+        _flush_status = s;
```
Comment (zhannngchen) [inline be/src/olap/memtable_flush_executor.cpp:150]: should return here?

File: be/src/olap/memtable_flush_executor.h [hunk 2]
```diff
@@ -87,7 +87,8 @@ class FlushToken {
 
     // Records the current flush status of the tablet.
     // Note: Once its value is set to Failed, it cannot return to SUCCESS.
-    std::atomic<int> _flush_status;
+    Status _flush_status;
+    std::shared_mutex _flush_status_lock;
```
Comment (zhannngchen) [inline be/src/olap/memtable_flush_executor.h:91]: It's normal put the lock in front of the variables it protects.

**Comment by bobhan1** [comment]
run buildall

**Review by zhannngchen — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**